scripts/events.js

```diff
@@ -0,0 +1,110 @@
+// Description
+//   Parses the UQCS calendar (https://calendar.google.com/calendar/ical/q3n3pce86072n9knt3pt65fhio%40group.calendar.google.com/public/basic.ics)
+//
+// Commands:
+//   !`events` _<description of time>_ - List UQCS Events
+//
+
+var ical = require('ical');
+module.exports = function (robot) {
+	robot.respond(/!?events ?(full|all|[1-9][0-9]*)? ?(weeks?)?/i, function (res) {
+		var months = ["JAN", "FEB", "MAR", "APR", "MAY", "JUN", "JUL", "AUG", "SEP", "OCT", "NOV", "DEC"];
+		var now = new Date().getTime();
+		var week = 1000*60*60*24*7;
+
+		var to = null; // Time until, false values imply no end
+		var cap = 0; // Maximum number of events, 0 implies no cap
+
+		var no_result_msg = "There doesn't appear to be any ";
+		var header;
+
+		if (res.match[1] === undefined) {
+			// Two weeks
+			to = now + week * 2;
+			no_result_msg += "events in the next two weeks...";
+			header = "Events in the *next two weeks*";
+		} else if (res.match[1] !== "full" && res.match[1] !== "all") {
+			var n = +res.match[1]; // Convert to num
+			if (res.match[2] === undefined) {
+				cap = n; // n events
+				no_result_msg += "upcoming events...";
+				header = "The *next _" + n + "_ events*";
+			} else {
+				to = now + week * n; // n weeks
+				no_result_msg += "events in the next *" + n + "* weeks...";
+				header = "Events in the *next _n_ weeks*";
+			}
+		} else {
+			// All
+			no_result_msg += "upcoming events...";
+			header = "List of *all* upcoming events";
+		}
+
+		ical.fromURL('https://calendar.google.com/calendar/ical/q3n3pce86072n9knt3pt65fhio%40group.calendar.google.com/public/basic.ics', {}, function(err, data) {
+			var arr = [];
+			for(var k in data) {
+				if(n === 0) { break; }
+				if(data[k].type == 'VTIMEZONE') { continue; } // I don't even know. See issue #50 on peterbraden/ical.js
+				if(data[k].rrule) {
+					// Repeated event, stored in the format from the rrule package
+					var times; // List of repeated times in the desired time-frame
+					if(to) {
+						times = data[k].rrule.between(new Date(now), new Date(to));
+					} else {
+						// Technically until a year ahead, safer this way.
+						times = data[k].rrule.between(new Date(now), new Date(now + week * 52));
+					}
+					var obj;
+					var duration = data[k].end.getTime() < data[k].start.getTime(); // How long it goes for
+					for(var i = 0; i < times.length; i++) {
+						obj = {};
+						// Copy necessary information from the original event
+						obj.location = data[k].location;
+						obj.summary = data[k].summary;
+						// Set new start/end
+						obj.start = times[i];
+						obj.end = new Date(times[i].getTime() + duration);
+
+						arr.push(obj);
+					}
+				} else if(data[k].end.getTime() > now) {
+					if(!to || data[k].start.getTime() < to) {
+						arr.push(data[k]);
+					}
+				}
+			}
+			arr.sort(function(a, b) {
+				return a.start.getTime() - b.start.getTime();
+			});
+
+			if(cap) {
```

Comment:
Well, this was here before adding recurring event support, and it is because if you just broke out of the loop when you had enough, you would have to sort before the loop. Seeing as the most common use-case only deals with a small percentage of the total number of events, iterating over them all, sorting, then truncating seemed like the nicest solution.

However, now that recurring events are also supported, this approach works even better. When you come across a recurring event, you can't know how many following events you care about (as it depends on how many other events there are etc); so adding them all, then truncating is a very clean solution in my opinion.